apollo-federation/src/query_plan/operation.rs

```diff
@@ -2515,8 +2534,119 @@ impl SelectionSet {
         self.selections.contains_key(key)
     }
 
-    pub(crate) fn add_at_path(&mut self, path: &OpPath, selection_set: Option<&Arc<SelectionSet>>) {
-        Arc::make_mut(&mut self.selections).add_at_path(path, selection_set)
+    fn add_selection(
+        &mut self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        schema: &ValidFederationSchema,
+        selection: Selection,
+    ) -> Result<(), FederationError> {
+        let selections = Arc::make_mut(&mut self.selections);
+
+        let key = selection.key();
+        let selection = match selections.remove(&key) {
+            None => selection,
+            Some(entry) => {
+                let to_merge = [entry, selection];
+                SelectionSet::make_selection(schema, parent_type, to_merge.iter())?.ok_or_else(
+                    || FederationError::internal("Failed to merge selections together."),
+                )?
+            }
+        };
+        /* TODO(@TylerBloom): Is this needed?
+        if key != selection.key() { todo!(); }
```

Comment:
My original goal was to use the `entry` method to get a reference to the selection inside the map and than pass a slice of references to selections to `make_selection` (but this didn't align with the proposed changes to `make_selection` as it needed a slice of owned values). From there, we could `std::mem::swap` the new selection with the existing one (if different). That was the rationale behind this check, but since we are removing the selection and then re-inserting it, I don't think this is needed.

Reply:
Thanks! 

The re-insert strategy changes the order of the selections unless we add more methods to `SelectionMap`. I was already thinking of adding a `SelectionMap::replace(key, |selection| -> selection)` API for other reasons, that may be the answer here.

Reply:
The `.replace()` idea isn't exactly what we need here since the key may not exist yet. For now I will use the existing remove + insert method but maintain the order.

Reply:
Oh, I saw your comment in between. Yes, a replace on `OccupiedEntry` could work